server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1140,11 +1141,23 @@ void runCleanup(ActionListener<DeleteResult> listener) {
         // Updating the shard-level metadata and accumulating results
 
         private void writeUpdatedShardMetadataAndComputeDeletes(ActionListener<Void> listener) {
-            try (var listeners = new RefCountingListener(listener)) {
-                for (IndexId indexId : originalRepositoryData.indicesToUpdateAfterRemovingSnapshot(snapshotIds)) {
-                    new IndexSnapshotsDeletion(indexId).run(listeners.acquire());
-                }
-            }
+            // noinspection resource -- closed safely at the end of the iteration
+            final var listeners = new RefCountingListener(listener);
+
+            // Each per-index process takes some nonzero amount of working memory to hold the relevant snapshot IDs and metadata generations
+            // etc. which we can keep under tighter limits and release sooner if we limit the number of concurrently processing indices.
+            // Each one needs at least one snapshot thread at all times, so threadPool.info(SNAPSHOT).getMax() of them at once is enough to
+            // keep the threadpool fully utilized.
```

Comment:
We discard the index metadata as soon as we've read it so I wouldn't expect that to be a big problem. I think the biggest consumer would be `survivingSnapshots`, a per-index copy of some subset of the total set of snapshot IDs.

I don't see a particular backpressure problem here but maybe I'm missing something. We're only executing one delete at once, maybe that's what you mean?

Reply:
Hmm I guess this is different from `staleBlobDeleteRunner` in terms of backpressure. The `ThrottledIterator` is a local variable while `staleBlobDeleteRunner` is an instance variable that is shared between all background cleanups. This means each `deleteSnapshots` creates its own `ThrottledIterator` so that the backpressure concern that we discussed last time does not apply here. Sorry for the false alarm.